proposed/2022/vulnerabilities-in-restore.md

```diff
@@ -0,0 +1,391 @@
+      }
+    ],
+    "2.3.0-beta100": [
+      {
+        "severity": 3,
+        "advisoryUrl": "https://github.com/advisories/GHSA-874w-m2v2-mj64"
+      }
+    ]
+  },
+  "Bond.Core.CSharp": {
+    "4.0.1": [
+      {
+        "severity": 2,
+        "advisoryUrl": "https://github.com/advisories/GHSA-rqrc-8q8f-cp9c"
+      }
+    ]
+  }
+} 
+```
+
+Having a dedicated vulnerability information resource will allow upstreams to easily use the NuGet.org data within their source.
+
+## Drawbacks
+
+- Vulnerability checking will have a performance hit.
+
+## Rationale and alternatives
+
+Most of the alternatives/ideas covered below are not exclusive.
+
+### Vulnerability dedicated resources updating alternatives
+
+There are many ways to solve the updating problem.
+
+Instead of using the `custom` @updated property, we could instead use HTTP constructs.
+
+- Use ETag instead of an intermediate resource. This would technically have 1 fewer call than the current proposal.
+- Use If-Modified-Since and simply always request both the base and update resources.
+- Version the vulnerabilities by the year informationw was known to source.
+  - Implement something similar to what the catalog is.
+  - This would allow certain vulnerabilities to never be redownloaded, but it will continously increased, even if it is unlikely to ever become unreasonable.
+
+- Multiple resource files organized in buckets such as the first character of the package id or some other [arbitrary bucketing method](https://en.wikipedia.org/wiki/Consistent_hashing).
+  - This will not improve the overall performance, but theoretically would limit the frequency of the data being refreshed.
+
+- Multiple resource files organized by severity. This could reduce the traffic and make restore faster, but only if customers only care about certain severities.
+
+### Vulnerability reporting not tied to sources
+
+All of NuGet's HTTP traffic is through sources.
+The vulnerability information does not have to be tied to sources.
+It could be something that's configured independently.
+
+This information could be something that's provided by nuget.org, similarly to licenses.nuget.org.
+This would be a url that the user would provide in their configuration. The advantage is that the user does not need their source to provide any vulnerability information.
+
+- Pros:
+  - Vulnerabilities are not inherently tied to sources. This is explicit effort undertaken by nuget as shepherds of the eco-system to provide vulnerability information.
+  - This can be additive. We can ship vulnerabilities tied to sources first and then update later.
+
+- Cons:
+  - The client has an established protocol and a means to configure a source.
+  - Experimentation is arguably easier as we're not introducing new syntax, at least not immediately.
+  - The challenge is that we'd need to establish a protocol, and something that could be slower on the uptake.
+
+### Optimizing vulnerability check cadence

```

Comment:
Would it make sense to update vulnerability info in the background in Visual Studio? That way, restore performance in Visual Studio wouldn't be affected by this feature.

Reply:
Good idea!  
Additionally this could be another step in CI (cli), so that not every build would be affected by a new download or in other words: 
Have more control over the time when a new file is going to be downloaded

Reply:
> Would it make sense to update vulnerability info in the background in Visual Studio? That way, restore performance in Visual Studio wouldn't be affected by this feature.

We don't expect this to be taking so long where it's noticeable consistently. 

On the topic of background downloading, it's not always desirable to communicate with sources wit hut intent. That could for example trigger authentication and we wouldn't want anything to interrupt the user. 

> Additionally this could be another step in CI (cli), so that not every build would be affected by a new download or in other words:

There are a lot of optimizations that happen prior to vulnerability data being downloaded. 
If you're not changing your packages/deleting the obj folder, restore will be snappy and won't redownload sources.